core/resource_manager.hpp

```diff
@@ -0,0 +1,168 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2023 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrei Lobov
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "shared.hpp"
+#include "utils/assert.hpp"
+#include "utils/misc.hpp"
+#include <memory>
+
+
+#if (defined(__clang__) || defined(_MSC_VER) || \
+     defined(__GNUC__) &&                       \
+       (__GNUC__ > 8))  // GCCs <=  don't have "<version>" header
+#include <version>
+#endif
+
+#ifdef __cpp_lib_memory_resource
+#include <memory_resource>
+#endif
+
+namespace irs {
+struct IResourceManager {
+  static IResourceManager kNoop;
+#ifdef IRESEARCH_DEBUG
+  static IResourceManager kForbidden;
+#endif
+
+  static_assert(sizeof(size_t) <= sizeof(uint64_t));
+
+  IRS_FORCE_INLINE bool Increase(size_t v) noexcept {
+    IRS_ASSERT(v <= static_cast<uint64_t>(std::numeric_limits<int64_t>::max()));
+    IRS_ASSERT(this != IResourceManager::kForbidden);
+    return Change(static_cast<int64_t>(v));
+  }
+
+  IRS_FORCE_INLINE void Decrease(size_t v) noexcept {
+    IRS_ASSERT(v <= static_cast<uint64_t>(std::numeric_limits<int64_t>::max()));
+    [[maybe_unused]] bool res = Change(-static_cast<int64_t>(v));
+    IRS_ASSERT(res);
+  }
+
+ protected:
+  virtual bool Change(int64_t) noexcept { return true; }
+};
+
+
+struct ResourceManagementOptions {
+
+  static ResourceManagementOptions kDefault;
+
+  IResourceManager& transactions{IResourceManager::kNoop};
+  IResourceManager& readers{IResourceManager::kNoop};
+  IResourceManager& consolidations{IResourceManager::kNoop};
+  IResourceManager& file_descriptors{IResourceManager::kNoop};
+  IResourceManager& cached_columns{IResourceManager::kNoop};
+};
+
+
+template<typename Allocator>
+class ManagedAllocator : private Allocator {
```

Comment:
no unique address works for msvc, ebo doesn't

Reply:
EBO works for MSVC. It is broken only for multiple inheritance.

Reply:
Anyway, no_unique_address sounds better for me